src/Features/Core/Diagnostics/EngineV1/DiagnosticAnalyzerDriver.cs

```diff
@@ -445,27 +445,55 @@ public bool IsAnalyzerSuppressed(DiagnosticAnalyzer analyzer)
 
-                // Get all the analyzer actions, including the per-compilation actions.
-                var analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
-                var analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
-                var hasDependentCompilationEndAction = await AnalyzerManager.Instance.GetAnalyzerHasDependentCompilationEndAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
-
-                if (hasDependentCompilationEndAction && forceAnalyzeAllDocuments != null)
+                try
                 {
-                    // Analyzer registered a compilation end action and at least one other analyzer action during its compilation start action.
-                    // We need to ensure that we have force analyzed all documents in this project for this analyzer before executing the end actions.
-                    forceAnalyzeAllDocuments(_project, analyzer, _cancellationToken);
-                }
+                    // Get all the analyzer actions, including the per-compilation actions.
+                    var analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
+                    var analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+                    var hasDependentCompilationEndAction = await AnalyzerManager.Instance.GetAnalyzerHasDependentCompilationEndAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+
+                    if (hasDependentCompilationEndAction && forceAnalyzeAllDocuments != null)
+                    {
+                        // Analyzer registered a compilation end action and at least one other analyzer action during its compilation start action.
+                        // We need to ensure that we have force analyzed all documents in this project for this analyzer before executing the end actions.
+                        // Doing so on the original analyzer instance might cause duplicate callbacks into the analyzer.
+                        // So we create a new instance of this analyzer and compute compilation diagnostics on that instance.
+
+                        try
+                        {
+                            analyzer = Activator.CreateInstance(analyzer.GetType()) as DiagnosticAnalyzer;
+                        }
+                        catch
+                        {
+                            // Unable to created a new analyzer instance, bail out on reporting diagnostics.
+                            return;
+                        }
+
+                        analyzerExecutor = GetAnalyzerExecutor(analyzer, compilation, localDiagnostics.Add);
+                        analyzerActions = await this.GetAnalyzerActionsAsync(analyzer, analyzerExecutor).ConfigureAwait(false);
+                        analyzerToDispose = analyzer;
```

Comment:
You might want to move this assignment to be immediately after the assignment to "analyzer" within the try above, to avoid risk of something that might throw an exception creeping in between that assignment and this one. (If that would not be correct, because ClearAnalyzerState requires the two statements above to execute first, then I suggest a comment to the effect that this is the earliest correct point for this assignment.)